index.d.ts

```diff
@@ -44,26 +45,24 @@ declare namespace snakecaseKeys {
   @template Path - Path of keys.
   */
   export type SnakeCaseKeys<
-    T extends Record<string, any> | readonly any[],
+    T extends ObjectOptional | readonly any[],
     Deep extends boolean = true,
     Exclude extends readonly unknown[] = EmptyTuple,
     Path extends string = ""
   > = T extends readonly any[]
     ? // Handle arrays or tuples.
       {
-        [P in keyof T]: T[P] extends Record<string, any> | readonly any[]
+        [P in keyof T]: T[P] extends Record<string, unknown> | readonly any[]
         ? SnakeCaseKeys<T[P], Deep, Exclude>
         : T[P];
       }
-    : T extends Record<string, Date | Error | RegExp>
-      ? T // Date, Error, and RegExp objects are not converted.
-      : T extends Record<string, any>
+    : T extends Record<string, unknown>
       ? // Handle objects.
         {
           [P in keyof T as [Includes<Exclude, P>] extends [true]
             ? P
             : SnakeCase<P>]: [Deep] extends [true]
-            ? T[P] extends Record<string, any> | undefined
+            ? T[P] extends ObjectOptional | readonly any[]
```

Comment:
If the value of object is type array and there is an object in that array, the key of object is added to make it a snake_cased type.
Without `readonly any[]`, the following test will not pass.

```ts
expectType<{ foo_bar: { foo_baz: { foo_bar: Point; }[]; }[]; }>(
  snakecaseKeys({ fooBar: [{ fooBaz: [{ fooBar: point }] }] }, { deep: true })
);
```